client/lib/src/server.rs

```diff
@@ -2,79 +2,73 @@
 
 use std;
 
-use common::socket::{SendSocket, ReceiveSocket};
+use std::sync::{Arc, Mutex};
+use std::sync::mpsc::{Sender, Receiver, TryRecvError};
+use std::sync::atomic::{Ordering, AtomicUsize};
 
-#[allow(missing_docs)]
-pub mod send {
-  use std;
-  use std::sync::mpsc::Sender;
+use bincode;
 
-  use common::protocol;
+use common::protocol;
+use common::socket::{SendSocket, ReceiveSocket};
 
-  #[derive(Clone)]
-  pub struct T {
-    pub sender     : Sender<Vec<u8>>,
-    pub bytes_sent : std::sync::Arc<std::sync::Mutex<u64>>,
-  }
+#[allow(missing_docs)]
+#[derive(Clone)]
+pub struct SSender {
+  // A boxed slice is used to reduce the sent size
+  pub sender: Sender<Box<[u8]>>,
+  // Please replace with AtomicU64 when it becomes stable
+  pub bytes_sent: Arc<AtomicUsize>,
+}
 
-  pub fn new(sender: Sender<Vec<u8>>) -> T {
-    T {
-      sender     : sender,
-      bytes_sent : std::sync::Arc::new(std::sync::Mutex::new(0)),
+impl SSender {
+  #[allow(missing_docs)]
+  pub fn new(sender: Sender<Box<[u8]>>) -> SSender {
+    SSender {
+      sender: sender,
+      bytes_sent: Arc::new(AtomicUsize::new(0)),
     }
   }
 
-  impl T {
-    pub fn tell(&self, msg: &protocol::ClientToServer) {
-      use bincode;
-      use bincode::serialize;
-      let msg = serialize(msg, bincode::Infinite).unwrap();
-      *self.bytes_sent.lock().unwrap() += msg.len() as u64;
-      self.sender.send(msg).unwrap();
-    }
+  #[allow(missing_docs)]
+  pub fn tell(&self, msg: &protocol::ClientToServer) {
+    let msg = bincode::serialize(msg, bincode::Infinite).unwrap();
+    // We aren't reading this until long after the write, so we use `Relaxed`
+    self.bytes_sent.fetch_add(msg.len() as usize, Ordering::Relaxed);
```

Comment:
Maybe I'm wrong, but it seems like if there are no reads or writes "near" this `fetch_add`, then `Relaxed` is the same as `SeqCst` and `SeqCst` is more obviously correct? I find it hard to properly reason about `Relaxed` orderings, maybe it's just me.

Reply:
According to http://en.cppreference.com/w/cpp/atomic/memory_order

> The following four requirements are guaranteed for all atomic operations: 
> (...)
> 4) Write-read coherence: if a side effect (a write) X on an atomic object M happens-before a value computation (a read) B of M, then the evaluation B shall take its value from X or from a side effect Y that follows X in the modification order of M

I chose it for the theoretical performance improvement over `SeqCst`. I can change it to `SeqCst` if you want.

Reply:
I'm not super familiar with memory ordering semantics. I think I'm convinced that this is fine. Thanks!